File: src/main/scala/top/scaleda/tcl/TclParserDefinition.java
package top.scaleda.tcl;

import com.intellij.lang.ASTNode;
import com.intellij.lang.ParserDefinition;
import com.intellij.lang.PsiParser;
import com.intellij.lexer.Lexer;
import com.intellij.openapi.project.Project;
import com.intellij.psi.FileViewProvider;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.intellij.psi.tree.IElementType;
import com.intellij.psi.tree.IFileElementType;
import com.intellij.psi.tree.TokenSet;
import org.antlr.intellij.adaptor.lexer.ANTLRLexerAdaptor;
import org.antlr.intellij.adaptor.lexer.PSIElementTypeFactory;
import org.antlr.intellij.adaptor.parser.ANTLRParserAdaptor;
import org.antlr.v4.runtime.Parser;
import org.antlr.v4.runtime.tree.ParseTree;
import org.jetbrains.annotations.NotNull;
import top.scaleda.tcl.parser.TclLexer;
import top.scaleda.tcl.parser.TclParser;
import top.scaleda.tcl.psi.factory.TclPsiNodeFactory;

public final class TclParserDefinition implements ParserDefinition {
  public TclParserDefinition() {
    PSIElementTypeFactory.defineLanguageIElementTypes(TclLanguage.INSTANCE(), TclParser.VOCABULARY, TclParser.ruleNames);
  }

  @Override
  public @NotNull Lexer createLexer(Project project) {
    return new ANTLRLexerAdaptor(TclLanguage.INSTANCE(), new TclLexer(null));
  }

  @Override
  public @NotNull PsiParser createParser(Project project) {
    return new ANTLRParserAdaptor(TclLanguage.INSTANCE(), new TclParser(null)) {
      @Override
      protected ParseTree parse(Parser parser, IElementType root) {
        return ((TclParser) parser).source_text();
      }
    };
  }

  @Override
  public @NotNull IFileElementType getFileNodeType() {
    return new IFileElementType(TclLanguage.INSTANCE());
  }

  @Override
  public @NotNull TokenSet getWhitespaceTokens() {
    return PSIElementTypeFactory.createTokenSet(
            TclLanguage.INSTANCE(),
            TclLexer.WhiteSpace
    );
  }

  @Override
  public @NotNull TokenSet getCommentTokens() {
    return PSIElementTypeFactory.createTokenSet(
            TclLanguage.INSTANCE(),
            TclLexer.OneLineComment
    );
  }

  @Override
  public @NotNull TokenSet getStringLiteralElements() {
    return PSIElementTypeFactory.createTokenSet(
            TclLanguage.INSTANCE(),
            TclLexer.String
    );
  }

  @Override
  public @NotNull SpaceRequirements spaceExistenceTypeBetweenTokens(ASTNode left, ASTNode right) {
    return SpaceRequirements.MAY;
  }

  @Override
  public @NotNull PsiFile createFile(@NotNull FileViewProvider viewProvider) {
    return new TclPSIFileRoot(viewProvider);
  }

  @Override
  public @NotNull PsiElement createElement(ASTNode node) {
    return TclPsiNodeFactory.create(node);
  }
}

File: src/main/scala/top/scaleda/tcl/TclParserDefinition.scala
package top.scaleda
package tcl

import tcl.parser.{TclLexer, TclParser}
import tcl.psi.factory.TclPsiNodeFactory

import com.intellij.lang.{ASTNode, ParserDefinition, PsiParser}
import com.intellij.lexer.Lexer
import com.intellij.openapi.project.Project
import com.intellij.psi.tree.{IElementType, IFileElementType}
import com.intellij.psi.{FileViewProvider, PsiFile}
import org.antlr.intellij.adaptor.lexer.{ANTLRLexerAdaptor, PSIElementTypeFactory}
import org.antlr.intellij.adaptor.parser.ANTLRParserAdaptor
import org.antlr.v4.runtime.Parser

final class TclParserDefinition extends ParserDefinition {
  override def createLexer(project: Project): Lexer =
    new ANTLRLexerAdaptor(TclLanguage.INSTANCE, new TclLexer(null))

  override def createParser(project: Project): PsiParser =
    new ANTLRParserAdaptor(TclLanguage.INSTANCE, new TclParser(null)) {
      override def parse(parser: Parser, root: IElementType) = {
        // TODO: check this
        parser.asInstanceOf[TclParser].inicio()
      }
    }

  override def getWhitespaceTokens = TclParserDefinition.WHITESPACE

  override def getFileNodeType = TclParserDefinition.FILE

  override def getCommentTokens = TclParserDefinition.COMMENTS

  override def getStringLiteralElements = TclParserDefinition.STRING

  override def createElement(node: ASTNode) = TclPsiNodeFactory.create(node)

  override def createFile(viewProvider: FileViewProvider): PsiFile =
    new TclPSIFileRoot(viewProvider)

  override def spaceExistenceTypeBetweenTokens(left: ASTNode, right: ASTNode) =
    ParserDefinition.SpaceRequirements.MAY
}

//noinspection DuplicatedCode
object TclParserDefinition {
  private val vocabulary = TclParser.VOCABULARY
//  val tokenNames = (0 to vocabulary.getMaxTokenType) // LONG RINGS THE ALARM BELL
//    .map(x => (x, vocabulary.getLiteralName(x)))
//    .map(x => (x, if (x._2 == null) vocabulary.getSymbolicName(x._1) else x._2))
//    .map(x => if (x._2 == null) "<INVALID>" else x._2)
//    .toArray
  PSIElementTypeFactory.defineLanguageIElementTypes(
    TclLanguage.INSTANCE,
    // tokenNames :+ "<INVALID>",
    vocabulary,
    TclParser.ruleNames
  )
  val FILE   = new IFileElementType(TclLanguage.INSTANCE)
  val tokens = PSIElementTypeFactory.getTokenIElementTypes(TclLanguage.INSTANCE)
  TclLogger.Log.info("tokens: " + tokens.toArray.mkString(", "))
  val COMMENTS = PSIElementTypeFactory.createTokenSet(
    TclLanguage.INSTANCE,
    TclLexer.COMMENT,
    TclLexer.COMMENT_INLINE
  )
  val WHITESPACE = PSIElementTypeFactory.createTokenSet(
    TclLanguage.INSTANCE,
    TclLexer.WS
  )
  val STRING = PSIElementTypeFactory.createTokenSet(
    TclLanguage.INSTANCE,
    TclLexer.CONST_STRING
  )
}
